src/mongocxx/test/spec/entity.hh

```diff
@@ -43,11 +42,28 @@ class map {
     map(map&&) = default;
     map& operator=(map&&) = default;
 
-    void insert(const key_type& key, mongocxx::client&& client);
+    ~map() = default;
 
-    bool contains(const key_type& key) const;
+    template <typename Entity>
+    bool insert(const key_type& key, Entity&& e) {
+        _map.emplace(key, std::forward<Entity>(e));
+        return _map.find(key) != std::end(_map);
+    }
+
+    bool insert(const key_type& key, client&& c);
+
+    client& get_client(const key_type& key);
+    database& get_database(const key_type& key);
+    collection& get_collection(const key_type& key);
 
    private:
+    // Objects are destroyed in reverse order of their appearance in the class definition. Since the
+    // client must outlive the objects created from it, the client objects are held in a separate
+    // map and declared first.
+    //
+    // @see: http://mongoc.org/libmongoc/current/lifecycle.html#object-lifecycle
+    // @see: https://isocpp.org/wiki/faq/dtors#order-dtors-for-members
+    std::unordered_map<key_type, client> _client_map;
```

Comment:
Interesting, and thanks for the context around order of member destruction. I don't see an obvious wrong order of destruction. If you do think there is a latent bug, it might be worth investigating with an individual test of the entity map.